### PR TITLE
Generator options

### PR DESCRIPTION
**Pull request by ms14981**
I'm not convinced by the file structure, so let me know if you have a more sensible idea!

### CODE REVIEW

File: src/commands/forgeCommand.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        const generatorOptions = Object.keys(options).filter((key) =>
+          key.startsWith(generatorOptionsPrefix)
+        );
+        generatorOptions.forEach((option) => {
+          const optionName = option.substring(generatorOptionsPrefix.length);
+          process.env[optionName] = options[option];
+        });
+
+        await generate(schema, generatorPath, options);
+        generatorResolver.cleanup();
```
Comment (ms14981) [inline src/commands/forgeCommand.js:55]: I moved this cleanup because it was happening before we'd finished with the  files

**Review by gkocak-scottlogic — APPROVED** [review]
👍

File: src/generatorOptions/index.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    .command("generator-options")
+    .description(
+      "List the options available for a generator. Some generators take additional options that configure their output, this command lists and describes the options available."
+    )
+    .argument(
+      "<generator>",
+      "Git URL, file path or npm package of a language-specific generator"
+    )
+    .action(async (generator) => {
+      log.standard(await generatorOptions(generator));
```
Comment (ColinEberhardt) [inline src/generatorOptions/index.js:15]: I don't think the use of `log.standard` makes sense here. Currently the logging module allows us to configure the verbosity of the forge command. In this instance, I can't see a reason to hide the generator options output, also this command doesn't support setting the verbosity either.

On that note, perhaps the log module lives in the `src/forge` folder?

File: src/forge/index.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+      "1"
+    )
+    .allowUnknownOption()
+    .action(async (schema, generatorPathOrUrl) => {
+      const generatorPath = generatorResolver.getGenerator(generatorPathOrUrl);
+      const configFile = path.join(generatorPath, "config.json");
+
+      // re-configure the command to generate the API
+      command.allowUnknownOption(false).action(async (_, __, options) => {
+        // set the additional options as environment variables
```
Comment (ColinEberhardt) [inline src/forge/index.js:45]: Thinking how these options might be used ...

Adding them as environment variables makes them accessible to helpers, which is how they're used in the [JavaScript generator](https://github.com/ScottLogic/openapi-forge-javascript/blob/main/helpers/import.js#L2). However, I can think of other instances where you might want to access them from the handlebars template.

I think it is worth adding them to the schema object, in the same way that 'standard' options are added:

https://github.com/ScottLogic/openapi-forge/blob/master/src/generate.js#L212

**Review by ColinEberhardt — CHANGES_REQUESTED** [review]
A couple of minor comments.

I'm quite happy with the proposed folder structure

File: src/forge/index.js (hunk 3 above)
Comment (ms14981) [reply src/forge/index.js:45]: It is already added. When running
```
./src/index.js forge https://petstore3.swagger.io/api/v3/openapi.json https://github.com/ScottLogic/openapi-forge-javascript.git -o js -l verbose --generator.moduleFormat "esmodule"
```
schema._options becomes

```
{
  "exclude": "",
  "output": "js",
  "logLevel": "verbose",
  "generator.moduleFormat": "esmodule"
}
```
Is this what you meant, or did you mean something different?

File: src/generatorOptions/index.js (hunk 2 above)
Comment (ms14981) [reply src/generatorOptions/index.js]: Changed the log line. Moved the log file to `src/common/`

File: src/forge/index.js (hunk 3 above)
Comment (ColinEberhardt) [reply src/forge/index.js:45]: Oh, perfect, that does the trick :-)

**Pull request merged**